### PR TITLE
Implement a generic Distribution.expand() and use it in a few places

### PR DESCRIPTION
**Pull request by fritzo**
This aims to make it easier to implement new distributions by providing a best-effort default implementation of `Distribution.expand()`, which is possible with the new `Constraint.event_dim` metadata provided by #50581 and #51369.  While this simple default implementation does not work in all cases, this PR aims to support the following common cases of new distributions (only some of which currently work; discussion needed):
- [x] All the distribution's data is stored as instance attributes described by `.arg_constraints` (e.g. Normal, Cauchy, Poisson);
- [ ] The distribution `.__init__()` simply constructs a `base_dist` and `tansforms` and passes those to `TransformedDistribution.__init__()`;
- [ ] The distribution inherits from another distribution and the `.__init__()` method simply changes parameters (e.g. Chi2)

To make this work, I've refactored some instance attributes such as `._param` and `._event_ndim` to be properties or lazy properties.

## Questions for reviewers
- [ ] Should we weaken the check in [Distribution._get_checked_instance()](https://github.com/pytorch/pytorch/blob/68d690ffbd64d0fb697dc3da1635216366649787/torch/distributions/distribution.py#L279)?
- [ ] How should we document and check whether the default `.expand()` works correctly?

cc @neerajprad @fehiepsi

### CODE REVIEW

File: torch/distributions/binomial.py [hunk 1]
```diff
@@ -50,19 +50,9 @@ def __init__(self, total_count=1, probs=None, logits=None, validate_args=None):
...
-            new.logits = self.logits.expand(batch_shape)
-            new._param = new.logits
-        super(Binomial, new).__init__(batch_shape, validate_args=False)
-        new._validate_args = self._validate_args
-        return new
+    @property
+    def _param(self):
+        return self.__dict__.get("probs", self.__dict__.get("logits"))
 
     def _new(self, *args, **kwargs):
```
Comment (neerajprad) [inline torch/distributions/binomial.py:57]: Not related to this PR but I'm just noticing this `_new` method now, and I am not sure if this is needed or used by anything. We can probably remove this in a separate PR.

**Comment by neerajprad** [comment]
> Should we weaken the check in Distribution._get_checked_instance()?

When do you think this will be useful? I think if we used something like `ExpandedDistribution` in Pyro as the default `.expand`, it would make sense to remove the class check.

File: torch/distributions/lkj_cholesky.py [hunk 2]
```diff
@@ -69,6 +69,7 @@ def __init__(self, dim, concentration=1., validate_args=None):
         super(LKJCholesky, self).__init__(batch_shape, event_shape, validate_args)
 
     def expand(self, batch_shape, _instance=None):
+        # This needs a custom .expand() because the dim arg is not in arg_constraints.
```
Comment (neerajprad) [inline torch/distributions/lkj_cholesky.py:72]: If it is simpler, I suppose we can add `dim` to `arg_constraints` but the scalar constraint on `dim` cannot be checked at the moment.

Comment (fritzo) [reply torch/distributions/lkj_cholesky.py:72]: Right, I think there is an implicit assumption on `Constraints` that they are batchable / broadcastable, but this `dim` param is a non-batchable constant.  I think the strictly proper way to pass `dim` would be as an array whose shape we read, e.g.
```py
class LKJCholesky(Distribution):
    arg_constraints = {"dim_prototype": constraints.independent(constraints.real, 1),
                       "concentration": constraints.positive}
    def __init__(self, dim_prototype, concentration, validate_args=None):
        self.dim = dim_prototype.size(-1)
        ...
```
but that seems awkwardly pedantic.

**Comment by fritzo** [comment]
> > Should we weaken the check in Distribution._get_checked_instance()?

> When do you think this will be useful?

Well I'd eventually like to improve usability by enabling two new use cases of custom `.__init__()` methods without custom `.expand()` methods, but I'm unsure whether they'll fit into this PR:
1. Subclass a concrete distribution class with special cases of parameters, e.g. `Chi2(df) = Gamma(df/2, 1/2)`.
2. Subclass `TransformedDistribution`, `Logistic(loc, scale) = TransformedDistribution(...)`.

File: torch/distributions/lkj_cholesky.py (hunk 2 above)
Comment (fehiepsi) [reply torch/distributions/lkj_cholesky.py:72]: So far, we have 3 types of args: tensors, constants (str, int,...), and base distributions. For expanding, I guess we can expand tensors and base distributions, and keep the constants as-is. The tricky thing is we currently don't have a mechanism to declare how arguments stored in a distribution instance. E.g. `dim` may be stored in attributes `.dim` or `._dim`. If we can make a restriction that `dim` is stored as `.dim`, then I guess we can inspect the `__init__` method and pass those constants to the expanded instance (alternatively, we can declare `dim` as a special constraint in `arg_constraints` as @neerajprad suggested, or create a new attribute named `extra_args`,...). I can see a few controversial arguments here.

Comment (fritzo) [reply torch/distributions/lkj_cholesky.py:72]: @fehiepsi interesting, I guess we could also consider `Transform`s as constants since they are not expanded. I guess we could add an additional piece of data, say `base_distributions : Frozenset[str]` that would work to automate most cases, with a couple weird cases like `IndependentDistribution` and `MixtureSameFamily`...

Comment (fehiepsi) [reply torch/distributions/lkj_cholesky.py:72]: Yes, I think `Transform` (and e.g. neural networks) can be considered as constants. About IndependentDistribution and MixtureSameFamily, I think we can check the base distribution's batch shape to see if its left-most dimensions match the main distribution's batch_shape. If not, we raise not ImplementedError. Otherwise, we perform expanding:
```python
def expand(self, batch_shape):
    # here base_dist can have other names
    new_base_dist = self.base_dist.expand(batch_shape + self.base_dist.batch_shape[len(self.batch_shape):])
```

For more complicated distributions, I guess we need to use custom `_expand` methods (e.g. to not expand base distributions and improve the performance - like in GaussianHMM).
